docs/develop-an-agent.md

```diff
@@ -0,0 +1,350 @@
+signature_data="$(printf '%s\n%s' "$resource_uri" "$expiry" | base64 -w 0)"
+```
+
+#### **Step 3** Get a signed token from the Keys Service
+
+Send a request to the Keys Service endpoint `/run/aziot/keyd.sock` to sign the base64-encoded value using the HMAC-SHA256 algorithm.
+
+If the connection is refused, go back to Step 1 since IoT Hub device information could have changed.
+
+Otherwise, parse the `signature` from the response.
+
+```bash
+# Parse and URL-decode the 'signature' from the response
+signature="$(
+    sudo curl --unix-socket '/run/aziot/keyd.sock' \
+        -X POST -H 'content-type: application/json' --data-raw "$(
+            jq -cn --arg 'keyHandle' "$key_handle" --arg 'message' "$signature_data" '{
+                "keyHandle": $keyHandle,
+                "algorithm": "HMAC-SHA256",
+                "parameters": {
+                    "message": $message,
+                },
+            }'
+        )" 'http://keyd.sock/sign?api-version=2020-09-01' |
+        jq '.signature' -r |
+        sed -e 's|+|%2b|g' -e 's|/|%2f|g' -e 's|=|%3d|g'
+)"
+
+# Construct the full SaS token
+sas_token="$(printf 'sr=%s&se=%s&sig=%s' "$resource_uri" "$expiry" "$signature")"
+```
+
+#### **Step 4**: Create the full connection string
+
+The device may have been configured to connect directly to IoT Hub, or it may be connecting via a gateway. It's direct if the `gatewayHost` obtained previously in **Step 1** is the same as the `hubName`. If they differ then connection string should use the gateway host name.
+
+```bash
+# Note that the 'SharedAccessSignature' appears both in the connection string and in the value for it.
+sas="SharedAccessSignature $sas_token"
+
+if [ "$hub_name" = "$gateway_host" ]; then
+    # Connecting directly to Hub.
+    echo "HostName=$hub_name;DeviceId=$device_id;ModuleId=$module_id;SharedAccessSignature=$sas"
+else
+    # Connecting via a gateway (e.g. in a nested configuration). Include the gateway name in the connection string.
+    echo "HostName=$hub_name;DeviceId=$device_id;ModuleId=$module_id;SharedAccessSignature=$sas;GatewayHost=$gateway_host"
+fi
+
+```
+
+#### **Step 5** Use the SDK to connect to IoT Hub
+
+For example, use the C SDK's [IoTHubDeviceClient_LL_CreateFromConnectionString](https://docs.microsoft.com/azure/iot-hub/iot-c-sdk-ref/iothub-device-client-ll-h/iothubdeviceclient-ll-createfromconnectionstring) API to create a client from the connection string.
+
+#### **Step 6** After disconnect
+
+Repeat all steps starting from _Step 1_ above to re-connect.
+
+### X.509 authentication for Modules
+
+Currently, host processes using module identities can only authenticate against IoT Hub using the SaS authentication method. 
+
+It is beneficial that you include and validate X.509 authentication in the short term (by registering a device principal) so that module X.509 authentication could be easily supported in the future. 
```

Comment:
```suggestion
It is beneficial that you include and validate X.509 authentication in the short term (by registering a device principal) so that module X.509 authentication can be easily supported in the future. 
```